server/src/main/java/org/elasticsearch/index/mapper/ArraySourceValueFetcher.java

```diff
@@ -38,7 +39,7 @@ public ArraySourceValueFetcher(String fieldName, SearchExecutionContext context)
      * @param nullValue A optional substitute value if the _source value is 'null'.
      */
     public ArraySourceValueFetcher(String fieldName, SearchExecutionContext context, Object nullValue) {
-        this.sourcePaths = context.sourcePath(fieldName);
+        this.sourcePaths = context.isSourceEnabled() ? context.sourcePath(fieldName) : Collections.emptySet();
```

Comment:
Could you explain the motivation for these checks `context.isSourceEnabled()`? Is it a small optimization to avoid extra work?

I saw that `FieldFetcherTests#testFetchFromSourceWithSourceDisabled` can fail without this check, but it seems like the test logic just needs to be adapted to not pass a null or empty `SourceLookup`?

Reply:
yes the check helps the new test work without providing a source lookup, but mostly I was annoyed that we do extra work when we can avoid it: why call sourcePath and iterate over what it returns when we already know that we have no _source?> And why even require a non-null SourceLookup in that case? Also, the way the test is currently setup, if I provide a non null _source, fields are going to be retrieved from it no matter whether source is disabled or not, because source is not retrieved from the index but rather provided manually to the fetch method.

I guess I took a shortcut to avoid having to deal with all of the above in this PR but I'd expect things to change once SearchLookup is no longer the only required argument. With this in mind this change did not feel so bad after all.

Reply:
++ thanks I will keep this in mind